File: backend/src/main/java/com/lsoft/aguapotable/controller/ComunidadController.java
package com.lsoft.aguapotable.controller;

import com.lsoft.aguapotable.domain.entity.Comunidad;
import com.lsoft.aguapotable.dto.request.ComunidadRequest;
import com.lsoft.aguapotable.service.ComunidadService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/v1/comunidades")
@RequiredArgsConstructor
@Tag(name = "Comunidades", description = "Gestión de comunidades (multi-tenant)")
@SecurityRequirement(name = "bearerAuth")
public class ComunidadController {

    private final ComunidadService comunidadService;

    @GetMapping
    @PreAuthorize("hasRole('SUPER_ADMIN')")
    @Operation(summary = "Listar todas las comunidades activas")
    public ResponseEntity<List<Comunidad>> listar() {
        return ResponseEntity.ok(comunidadService.listarActivas());
    }

    @GetMapping("/{id}")
    @Operation(summary = "Obtener comunidad por ID")
    public ResponseEntity<Comunidad> getById(@PathVariable Long id) {
        return ResponseEntity.ok(comunidadService.obtenerPorId(id));
    }

    @GetMapping("/mi-comunidad")
    @Operation(summary = "Obtener datos de la comunidad del usuario actual")
    public ResponseEntity<Comunidad> getMiComunidad(HttpServletRequest request) {
        Long comunidadId = (Long) request.getAttribute("comunidadId");
        return ResponseEntity.ok(comunidadService.obtenerPorId(comunidadId));
    }

    @PostMapping
    @PreAuthorize("hasRole('SUPER_ADMIN')")
    @Operation(summary = "Crear nueva comunidad")
    public ResponseEntity<Comunidad> crear(@Valid @RequestBody ComunidadRequest dto) {
        return ResponseEntity.status(201).body(comunidadService.crear(dto));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasAnyRole('SUPER_ADMIN','ADMIN')")
    @Operation(summary = "Actualizar datos de la comunidad")
    public ResponseEntity<Comunidad> actualizar(@PathVariable Long id,
                                                 @Valid @RequestBody ComunidadRequest dto) {
        return ResponseEntity.ok(comunidadService.actualizar(id, dto));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('SUPER_ADMIN')")
    @Operation(summary = "Desactivar comunidad")
    public ResponseEntity<Void> desactivar(@PathVariable Long id) {
        comunidadService.desactivar(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/comunidad/comunidad.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  ParseIntPipe,
  HttpCode,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiBearerAuth } from '@nestjs/swagger';
import { ComunidadService } from './comunidad.service';
import { ComunidadRequest } from './dto/comunidad.dto';
import { Comunidad } from '../common/entities';
import { ComunidadId } from '../common/decorators';
import { RolesGuard, Roles } from '../auth/guards/roles.guard';

@ApiTags('Comunidades')
@ApiBearerAuth()
@Controller('api/v1/comunidades')
@UseGuards(AuthGuard('jwt'), RolesGuard)
export class ComunidadController {
  constructor(private readonly comunidadService: ComunidadService) {}

  @Get()
  @Roles('SUPER_ADMIN')
  @ApiOperation({ summary: 'Listar todas las comunidades activas' })
  async listar(): Promise<Comunidad[]> {
    return this.comunidadService.listarActivas();
  }

  @Get('mi-comunidad')
  @ApiOperation({ summary: 'Obtener datos de la comunidad del usuario actual' })
  async getMiComunidad(@ComunidadId() comunidadId: number): Promise<Comunidad> {
    return this.comunidadService.obtenerPorId(comunidadId);
  }

  @Get(':id')
  @ApiOperation({ summary: 'Obtener comunidad por ID' })
  async getById(@Param('id', ParseIntPipe) id: number): Promise<Comunidad> {
    return this.comunidadService.obtenerPorId(id);
  }

  @Post()
  @Roles('SUPER_ADMIN')
  @HttpCode(201)
  @ApiOperation({ summary: 'Crear nueva comunidad' })
  async crear(@Body() dto: ComunidadRequest): Promise<Comunidad> {
    return this.comunidadService.crear(dto);
  }

  @Put(':id')
  @Roles('SUPER_ADMIN', 'ADMIN')
  @ApiOperation({ summary: 'Actualizar datos de la comunidad' })
  async actualizar(
    @Param('id', ParseIntPipe) id: number,
    @Body() dto: ComunidadRequest,
  ): Promise<Comunidad> {
    return this.comunidadService.actualizar(id, dto);
  }

  @Delete(':id')
  @Roles('SUPER_ADMIN')
  @HttpCode(204)
  @ApiOperation({ summary: 'Desactivar comunidad' })
  async desactivar(@Param('id', ParseIntPipe) id: number): Promise<void> {
    return this.comunidadService.desactivar(id);
  }
}
